### ISSUE TITLE
[ Debugging ] Natvis treats 'using' types as new types

### ISSUE THREAD

**Issue by iwubcode** [Microsoft/VSLinux#221]
**Problem Description**

When trying to debug a unit test.  It is difficult to view the raw information for some variables.  These variable types are technically stl types but they are defined as 'different' types with a _using_ statement.

**Expected Results**

Natvis should allow me to properly view my variables that are really stl types.

**Actual Results**

I have to define separate natvis rules to view these types.  Since our teams use 'using' almost everywhere, this can be quite cumbersome.

**Environment**
Visual Studio 2017 15.5 Preview 3
GDB 8.x

**How to reproduce**

* Create a project with a single class.
* Define a variable in that class with a type 'MyString'
* Define 'MyString' as a std::string
* Make an instance of that class and notice that you can't easily view the variable with the natvis shipping with VisualStudio

**Comment by paulmaybee** [comment]
This is an unfortunate result of using gdb's MI interface. The expression for a variable declared as a 'MyString' is reported with type 'MyString' and no additional information is available via the interface.

**Comment by iwubcode** [comment]
@paulmaybee - is there a workaround?  Pretty-printers seem to have its own issue or rather, they show all the _M_p internal stuff as well.  See my comment on #223.  Not sure if there's a bug with pretty-printing or not?

Anyway, if this Natvis issue is out of our hands, I'm happy to close this.

**Comment by paulmaybee** [comment]
You should not be seeing _M_p stuff with pretty-printing turned on. How did you do it from your CMake project? Here is an example of a CMake config that turns on pretty printing:

    {
      "type": "cppdbg",
      "name": "App_HelloWorld (examples\\HelloWorld\\App_HelloWorld)",
      "project": "CMakeLists.txt",
      "projectTarget": "App_HelloWorld (examples\\HelloWorld\\App_HelloWorld)",
      "cwd": "${debugInfo.remoteWorkspaceRoot}",
      "program": "${debugInfo.fullTargetPath}",
      "MIMode": "gdb",
      "externalConsole": true,
      "remoteMachineName": "${debugInfo.remoteMachineName}",
      "pipeTransport": {
        "pipeProgram": "${debugInfo.shellexecPath}",
        "pipeArgs": [
          "/s",
          "${debugInfo.remoteMachineId}",
          "/p",
          "${debugInfo.parentProcessId}",
          "/c"
        ]
      },
      "setupCommands": [
        {
          "text": "-enable-pretty-printing",
          "ignoreFailures": true
        }
      ]
    },

**Comment by iwubcode** [comment]
@paulmaybee - thanks!  I'll try this shortly.  Before, I was executing pretty-printing directly.

**Comment by iwubcode** [comment]
@paulmaybee - hmm, I'm not seeing a change.  Should I try and make a simple repo of the issue?  I was assuming pretty-printers would have the same issue as natvis, since gdb wasn't providing the actual type?

**Comment by paulmaybee** [comment]
Yes, send me the repro. Here is what I see when pretty printing is enabled on Ubuntu 16.04 with gcc 6.3.0

![image](https://user-images.githubusercontent.com/12514776/32741761-2f59fabe-c85c-11e7-9864-907a09257902.png)

**Comment by iwubcode** [comment]
@paulmaybee - do you still see it correctly if you do:

```
using MyString = std::string;

std::vector<MyString> foo(3);
foo[1] = "foo";
foo[2] = "bar";
MyString = bar = "barbar";
foo[0] = bar;
std::map<int, MyString> theMap;
theMap[1] = bar;
theMap[2] = foo[2];
int i = 0;
///------initialization_start------
while (i == 0);
```

**Comment by paulmaybee** [comment]
Yes.
![image](https://user-images.githubusercontent.com/12514776/32742072-2356807e-c85d-11e7-80e0-67742aedd73b.png)

**Comment by iwubcode** [comment]
Hi @paulmaybee , sorry for taking so long to get back to you.  I just now have had a chance to test.

In my same project, I added a simple vector of ints with both the type and the container type being defined in a 'using' statement.  Even with 'pretty print' I get a weird output.  Code looks like:

```
using MyInt = int;
using MyList = std::vector<MyInt>;
MyList test;
test.push_back(1);
test.push_back(2);
test.push_back(3);
```

Which results in:

![image](https://user-images.githubusercontent.com/15224722/35242102-417e461a-ff7e-11e7-8736-b99adb5f4b30.png)

While changing back test to a vector of ints seems to work correctly:

![image](https://user-images.githubusercontent.com/15224722/35242235-99a37414-ff7e-11e7-8e46-4a0a1740b232.png)


Are you able to see a similar issue?

**Comment by paulmaybee** [comment]
The vector formatting you are seeing is from natvis, not from gdb's pretty-printing. If pretty-printing is enabled it should take precedence over natvis formatting. 

That said, you can always add additional natvis files to your project. In order to accurately display your renames stl container content do the following:

1. In your solution right click and select "Add>New Item>Visual C++>Utility>Debugger visualization file. Call it something like "mytypes.natvis".

2. Find the  stl.natvis file that comes with the VS installation (at <your VS installation>\Common7\IDE\CommonExtensions\Microsoft\Linux\Linux\stl.natvis).

3. For each of your aliased stl types, copy the natvis entry from stl.natvis to mytypes.natvis. Change the name of the type to match your name. For example, for your type MyList defined as "vector<int>" find the vector entry:

  
  <Type Name="std::vector&lt;*&gt;">
    <DisplayString>{{size={_M_impl._M_finish - _M_impl._M_start}}}</DisplayString>
    <Expand>
      <Item Name="[size]" ExcludeView="simple">_M_impl._M_finish - _M_impl._M_start</Item>
      <Item Name="[capacity]" ExcludeView="simple">_M_impl._M_end_of_storage - _M_impl._M_start</Item>
      <ArrayItems>
        <Size>_M_impl._M_finish - _M_impl._M_start</Size>
        <ValuePointer>_M_impl._M_start</ValuePointer>
      </ArrayItems>
    </Expand>
  </Type>
 
and place it in your mytypes.natvis looking like:

  
  <Type Name="Mylist">
    <DisplayString>{{size={_M_impl._M_finish - _M_impl._M_start}}}</DisplayString>
    <Expand>
      <Item Name="[size]" ExcludeView="simple">_M_impl._M_finish - _M_impl._M_start</Item>
      <Item Name="[capacity]" ExcludeView="simple">_M_impl._M_end_of_storage - _M_impl._M_start</Item>
      <ArrayItems>
        <Size>_M_impl._M_finish - _M_impl._M_start</Size>
        <ValuePointer>_M_impl._M_start</ValuePointer>
      </ArrayItems>
    </Expand>
  </Type>

**Comment by iwubcode** [comment]
@paulmaybee - thanks.  I do know about that but we have quite a number of these special containers since the developers like to use them a lot... I'm happy to do that as a last resort but would prefer to avoid natvis if I can help it due to this issue.

I've got the following settings in my launch.vs.json file:

```
{
      "type": "cppdbg",
      "name": "MyProg-UT(bin\\MyProg-UT)",
      "project": "CMakeLists.txt",
      "projectTarget": "MyProg-UT (bin\\MyProg-UT)",
      "cwd": "${debugInfo.remoteWorkspaceRoot}",
      "program": "${debugInfo.fullTargetPath}",
      "MIMode": "gdb",
      "externalConsole": true,
      "remoteMachineName": "${debugInfo.remoteMachineName}",
      "pipeTransport": {
        "pipeProgram": "${debugInfo.shellexecPath}",
        "pipeArgs": [
          "/s",
          "${debugInfo.remoteMachineId}",
          "/p",
          "${debugInfo.parentProcessId}",
          "/c"
        ],
        "debuggerPath": "/usr/local/bin/gdb"
      },
      "setupCommands": [
        {
          "text": "-enable-pretty-printing",
          "ignoreFailures": true
        }
      ]
    }
```

Specifically the 'setupcommands', is there anything else I need to do to enable pretty-printing?

**Comment by paulmaybee** [comment]
That setup command should be all you need. Did you build your own gdb? Is it built with python pretty printing enabled? If you just use gdb locally on the Linux machine are you seeing the print command format your containers correctly? On my Ubuntu machine I see 
**(gdb)** p list
$1 = std::vector of length 2, capacity 2 = {0, 2}

when stopped at the printf in this program:

```
using MyList = std::vector<int>;

int main(int argc, char** argv)
{
	MyList list;
	list.push_back(0);
	list.push_back(2);
	printf("hello\n");
}
```

**Comment by iwubcode** [comment]
@paulmaybee - Sorry for getting back to you so late.  I was investigating this a bit.  I still don't have an explanation for why I'm seeing garbled information but I know it's based around the gdb that I was using because I see the same thing in gdb on the box.

I can say that we have multiple gdb on the box I am building on.  One is version 8.x and one is version 7.x.  The 8.x version outputs the garbled data that I mentioned earlier.

Upon testing the 7.x version, I found that it works for the simple use-case we were looking at.  However, as a separate issue it gives _"Children can not be evaluated"_ in some more complex scenarios (ex: trying to look at the details of a std shared pointer or unique pointer).  I wasn't sure how to access it in gdb, I tried printing it but it just prints the fact it's a shared_ptr.  So, I feel like this might be unrelated to Visual Studio and more a problem with gdb (or my gdb) itself.

My apologies for having this lengthy conversation in a (mostly) unrelated defect about natvis.  But thank you for all your help!
